server/database.js

```diff
@@ -133,64 +136,115 @@ function transformDbObjects(key, value) {
                 }
+            if (value instanceof Set)
+                return { dbSerializerTransform: 'set', values: Array.from(value) }
             else
                 return value
         default:
             return value
     }
 }
 
-function restoreDbObjects(db) {
+function restoreDbObjects() {
     for (const [id, obj] of Object.entries(db)) {
         // Everything but the 'refs' and 'tasks' fields should be a dbObject
         if (obj.dbSerializerTransform === 'dbObject') {
             // Transform function code back to functions
             for (const [field, value] of Object.entries(obj)) {
-                if (value.dbSerializerTransform === 'function')
+                if (value?.dbSerializerTransform === 'function')
                     obj[field] = new Function(`return ${value.code}`)()
             }
             // Set the parent object
             if (obj.parent) {
-                Object.setPrototypeOf(db[id], db[obj.parent])
-                delete db[id].parent
+                // Parent info will be an untransformed ref at this point.  It will get resolved later, but at this
+                // stage we need to dereference it manually
+                Object.setPrototypeOf(db[id], db[obj.parent.id])
             } else {
                 Object.setPrototypeOf(db[id], rootParent) // This is the root object
             }
             delete obj.dbSerializerTransform
             db[id] = obj // obj will have constructor.name === 'DatabaseObject'
         }
     }
-    restoreRefs(db, db)
+    restoreSets(db, [])
+    restoreRefs(db, [])
     restoreTasks(db)
 }
 
-function restoreRefs(node, db) {
-    if (typeof node !== 'object') return
+function restoreRefs(node, visited) {
+    if (typeof node !== 'object' || node === null || visited.includes(node)) return
+    visited.push(node)
     if (Array.isArray(node)) {
-        for (const entry of node) {
-            restoreRefs(entry, db)
+        for (const [index, entry] of node.entries()) {
+            if (entry?.dbSerializerTransform === 'ref') {
+                node[index] = db[entry.id]
+                // Since this was a ref to a DB object its props will be traversed elsewhere, we don't need to recurse
+            } else {
+                restoreRefs(entry, visited)
+            }
+        }
+        return
+    }
+    if (node instanceof Set) {
+        for (const member of new Set(node)) { // copy to avoid changing the set while we're iterating it
+            if (member?.dbSerializerTransform === 'ref') {
+                node.delete(member)
+                node.add(db[member.id])
+            } else {
+                restoreRefs(member, visited)
+            }
         }
         return
     }
+    // Regular objects
     for (const [key, value] of Object.entries(node)) {
-        if (value.dbSerializerTransform === 'ref') {
+        if (value?.dbSerializerTransform === 'ref') {
```

Comment:
I am kinda curious why these could be null now, if you want to explain in meatspace possibly over a rum and coke.

Reply:
It's just because the root object has a null parent.  So it's either this or a special check for that situation, and I'm a fan of brevity.

Reply:
But now that I think about it, I suppose there might be good reasons to have other props with null values.